recipes/full_finetune_distributed.py

```diff
@@ -502,6 +512,27 @@ def _setup_model(
         if self._compile:
             training.compile_model(model, verbose=self._is_rank_zero)
 
+        self.device_mesh = dist.init_device_mesh(
```

Comment:
I would rather explicitly save what we need here. So instead of putting the whole device_mesh on the object, later on if data_parallel_dim > 1, save self.local_size = dp_mesh.size() and self.local_rank = dp_mesh.get_local_rank() else self.local_size = self.world_size and self.local_rank = self.rank.

Does that make sense?